src/V3Const.cpp

```diff
@@ -2446,6 +2738,10 @@ class ConstVisitor final : public AstNVisitor {
     TREEOPV("AstRedAnd{$lhsp.castExtend, $lhsp->width() > VN_CAST($lhsp,,Extend)->lhsp()->width()}", "replaceZero(nodep)");  // &{0,...} => 0  Prevents compiler limited range error
     TREEOPV("AstRedOr {$lhsp.castExtend}",      "AstRedOr {$lhsp->castExtend()->lhsp()}");
     TREEOPV("AstRedXor{$lhsp.castExtend}",      "AstRedXor{$lhsp->castExtend()->lhsp()}");
+    TREEOP ("AstRedXor{$lhsp.castXor, VN_IS(VN_CAST($lhsp,,Xor)->lhsp(),,Const)}", "AstXor{AstRedXor{$lhsp->castXor()->lhsp()}, AstRedXor{$lhsp->castXor()->rhsp()}}");  // ^{const, b} => (^const)^(^b)
+    TREEOP ("AstRedAnd {$lhsp.castConcat, matchBitOpTree(nodep)}", "DONE");
+    TREEOP ("AstRedOr  {$lhsp.castConcat, matchBitOpTree(nodep)}", "DONE");
+    TREEOP ("AstRedXor {$lhsp.castConcat, matchBitOpTree(nodep)}", "DONE");
```

Comment:
> let V3Expand do its job first.

Let me confirm two things.
## 1
Do you mean `matchBitOptTree()` should be called in `constifyCpp()` that is after `expandAll()`
and   `constifyAll()` that is before `expandAll()` should not call `matchBitOpTree()` ?
https://github.com/verilator/verilator/blob/3c849d7ce0e70e95c30c32446e1b5372a12c3607/src/Verilator.cpp#L446-L454

## 2
Is it enough for letting V3Expand do first  to change from `TREEOP` to`TREEOPC`  ?
Or should `matchBitOpTree()` have a check something like below ?
```c++
    bool matchBitOpTree(AstNode* nodep) {
       if (!m_doCpp) return false;
```

Reply:
Ahh, you need a !m_doCpp.  I will rename TREEOPC to something else to avoid future trouble.

Reply:
I see. I will change to 
```c++
TREEOP("AstRedAnd {m_doCpp, $lhsp.castConcat, matchBitOpTree(nodep)}", "DONE");
TREEOP("AstRedOr  {m_doCpp, $lhsp.castConcat, matchBitOpTree(nodep)}", "DONE");
TREEOP("AstRedXor {m_doCpp, $lhsp.castConcat, matchBitOpTree(nodep)}", "DONE");
```
and 
```c++
TREEOP("AstAnd {m_doCpp, matchBitOpTree(nodep)}",           "DONE");
TREEOP("AstOr  {m_doCpp, matchBitOpTree(nodep)}",           "DONE");
TREEOP("AstXor {m_doCpp, matchBitOpTree(nodep)}",           "DONE");
```

Or do you prefer TREEOPA rather than TREEOP ?

Reply:
I think TREEOP (no A) with m_doCpp.  (Not !m_doCpp as had above).  Optionally change astgen to use now freed letter C to do the m_doCpp for you.

Reply:
Hmm...
After V3Expand, AstSel does not exist. It looks And(1, ShiftR(VarRef, lsb)).
I will do some experiment to see if enough information can be extracted from the representation.